ai/core/src/unitycatalog/ai/core/executor/local_subprocess.py

```diff
@@ -0,0 +1,154 @@
+import ast
+import base64
+import json
+import logging
+import os
+import subprocess
+import sys
+import tempfile
+import textwrap
+from typing import Any
+
+import cloudpickle
+
+from unitycatalog.ai.core.envs.executor_env_vars import (
+    EXECUTOR_MAX_CPU_TIME_LIMIT,
+    EXECUTOR_MAX_MEMORY_LIMIT,
+    EXECUTOR_TIMEOUT,
+)
+from unitycatalog.ai.core.executor.common import (
+    MB_CONVERSION,
+    NO_OUTPUT_MESSAGE,
+    TIMEOUT_ERROR_MESSAGE,
+)
+
+_logger = logging.getLogger(__name__)
+
+
+def _extract_function_name(function_source: str) -> str:
+    """
+    Extracts the name of the first function defined in the provided source code
+    using the ast module.
+    """
+    try:
+        module_ast = ast.parse(textwrap.dedent(function_source))
+    except SyntaxError as e:
+        raise ValueError("Invalid Python source code") from e
+
+    for node in module_ast.body:
+        if isinstance(node, ast.FunctionDef):
+            return node.name
+    raise ValueError("No function definition found in the provided source code.")
+
+
+def _generate_runner_script(function_source: str, cpu_time_limit: int, memory_limit: int) -> str:
+    """
+    Generates a Python runner script that sets resource limits and embeds the function's source code
+    directly into the script that will be executed in a subprocess.
+    """
+    function_name = _extract_function_name(function_source)
+    function_source = textwrap.dedent(function_source)
+    script = f"""
+import sys, base64, json, traceback, asyncio, resource, os
+
+{function_source}
+
+def _limit_resources():
+    try:
+        resource.setrlimit(resource.RLIMIT_CPU, ({cpu_time_limit}, {cpu_time_limit}))
+    except Exception as e:
+        sys.stderr.write("Warning: unable to set RLIMIT_CPU: " + str(e) + "\\n")
+    try:
+        mem_bytes = {memory_limit} * {MB_CONVERSION}
+        resource.setrlimit(resource.RLIMIT_AS, (mem_bytes, mem_bytes))
+    except Exception as e:
+        sys.stderr.write("Warning: unable to set RLIMIT_AS: " + str(e) + "\\n")
+
+def main():
+    _limit_resources()
+    try:
+        b64_params = sys.argv[1]
+        params = __import__('cloudpickle').loads(__import__('base64').b64decode(b64_params))
+        result = {function_name}(**params)
+        if asyncio.iscoroutine(result):
+            result = asyncio.run(result)
+        if result is None:
+            result = "{NO_OUTPUT_MESSAGE}"
+        print(__import__('json').dumps({{"success": True, "result": result}}))
```

Comment:
Why not `json.dumps` since you've already imported json? Similar for traceback below

Reply:
Yeah that's a good point. They're all available directly in the subprocess environment. I'll simplify that script :D